src/stan/error_handling/matrix/check_std_vector_index.hpp

```diff
@@ -11,28 +12,33 @@ namespace stan {
     /**
      * Return <code>true</code> if the specified index is valid in std vector
      *
+     * This check is 1-indexed by default. This behavior can be changed
+     * by setting <code>stan::error_index::value</code>.
+     *
      * NOTE: this will not throw if y contains nan values.
      *
-     * @param function
-     * @param i is index
-     * @param y std vector to test against
-     * @param name
+     * @tparam T Scalar type
+     *
+     * @param function Function name (for error messages)
+     * @param name Variable name (for error messages)
+     * @param y <code>std::vector</code> to test
+     * @param i Index
+     * 
      * @return <code>true</code> if the index is a valid in std vector.
-     * @tparam T Type of scalar.
+     * @throw <code>std::out_of_range</code> if the index is out of range.
      */
-    template <typename T_y>
+    template <typename T>
     inline bool check_std_vector_index(const std::string& function,
                                        const std::string& name,
-                                       const std::vector<T_y>& y,
+                                       const std::vector<T>& y,
                                        size_t i) {
-      if ((i > 0) && (i <= static_cast<size_t>(y.size())))
+      if ((i >= stan::error_index::value) 
+          && (i < y.size() + stan::error_index::value))
```

Comment:
My preference is to go with the language's built-in precedence and drop the parens around the comparisons.